runtime/lua/vim/net.lua

```diff
@@ -0,0 +1,53 @@
+local M = {}
+
+--- Downloads a file from the given URL to a specified location.
+---
+--- @param url string The URL to download from.
+--- @param download_location string The file path where the content should be saved.
+--- @param opts? table Optional parameters:
+---   - on_exit (function|nil): Called when the download completes or fails. Receives error string or nil.
+---   - verbose (boolean|nil): Enables verbose curl output.
+---   - retry (integer|nil): Number of times to retry on transient failures (default: 3).
+---
+--- @return nil
+function M.download(url, download_location, opts)
+  vim.validate({
+    url = { url, 'string' },
+    download_location = { download_location, 'string' },
+    opts = { opts, 'table', true },
+  })
+
+  opts = opts or {}
+  local retry = opts.retry or 3
+
+  local args = { 'curl' }
+
+  if opts.verbose then
+    table.insert(args, '--verbose')
+  else
+    vim.list_extend(args, { '--silent', '--show-error', '--fail' })
+  end
+
+  vim.list_extend(args, {
+    '--retry',
+    tostring(retry),
+    '--location',
+    '--output',
+    download_location,
+    url,
+  })
+
+  vim.system(args, { text = true }, function(res)
+    if res.code ~= 0 or res.stderr ~= '' then
+      if opts.on_exit then
+        opts.on_exit(res.stderr ~= '' and res.stderr or 'Download failed')
+      end
+      return
+    end
+    if opts.on_exit then
+      opts.on_exit(nil)
+    end
+  end)
```

Comment:
Perhaps add a sync form.

```suggestion
  local result

  vim.system(args, { text = true }, function(res)
    result = res
    if res.code ~= 0 or res.stderr ~= '' then
      if opts.on_exit then
        opts.on_exit(res.stderr ~= '' and res.stderr or 'Download failed')
      end
      return
    end
    if opts.on_exit then
      opts.on_exit(nil)
    end
  end)
  
  if on_exit then
    return
  end
  
  vim.wait(TIMEOUT, function()
    return result ~= nil
  end)
  
  if result.code ~= 0 then
    return nil, result.stdout or 'Download failed'
  end
  
  return result.stdout
```